library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -1178,17 +1182,41 @@ impl<T, A: Allocator> VecDeque<T, A> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn is_empty(&self) -> bool {
-        self.tail == self.head
+        self.len == 0
     }
 
-    fn range_tail_head<R>(&self, range: R) -> (usize, usize)
+    /// Given a range into the logical buffer of the deque, this function
+    /// return two ranges into the physical buffer that correspond to
+    /// the given range.
+    fn slice_ranges<R>(&self, range: R) -> (Range<usize>, Range<usize>)
     where
         R: RangeBounds<usize>,
     {
-        let Range { start, end } = slice::range(range, ..self.len());
-        let tail = self.wrap_add(self.tail, start);
-        let head = self.wrap_add(self.tail, end);
-        (tail, head)
+        let Range { start, end } = slice::range(range, ..self.len);
+        let len = end - start;
+
+        if len == 0 {
+            (0..0, 0..0)
```

Comment:
Is handling `len == 0` separately needed here?  It looks like the `head_len >= len` branch below would work fine with `start == end`, so my instinct is that avoiding the extra branch is better than saving an addition or two.

For example, this implementation means two branches in `as_slices`, I think, when it could just be the one necessary one there is today -- `.slice_ranges(..)` ends up at `start = 0`, so `self.to_physical_idx(start)` presumably trivially optimizes down to just `self.head`, removing that check, so there'd just be the "is this contiguous" check left.

Reply:
The problem I had was that `deque.slice_ranges(deque.len..)` would trigger the new debug assert in `wrap_index` if `deque.len == `deque.capacity()`, because technically the wrapping behavior in that case is wrong as the resulting index points outside of the buffer. In practice, it doesn't make a difference as the returned range would be 0 length anyways, but I just added the extra branch here to avoid having to specify the behavior of `self.to_physical_idx(self.capacity())`.

Reply:
Oh, interesting!  I guess this is the classic index-vs-fencepost dilemma, which is also why the `debug_assert` in `wrap_index` needs to allow `logical_index == 0 && capacity == 0`.

A thorough change would be to have `wrap_index` and `wrap_fencepost`, but let's not do that right now.  We can do little μoptimization tweaks as separable follow-ups.